hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mapreduce/Import.java

```diff
@@ -135,9 +135,7 @@ public CellWritableComparable(Cell kv) {
 
     @Override
     public void write(DataOutput out) throws IOException {
-      out.writeInt(PrivateCellUtil.estimatedSerializedSizeOfKey(kv));
-      out.writeInt(0);
-      PrivateCellUtil.writeFlatKey(kv, out);
+        KeyValueUtil.write(new KeyValue(kv), out);
```

Comment:
`PrivateCellUtil.writeFlatKey` writes the key value in a different format from `KeyValueUtil.write`. This breaks compatibility, can we just keep writing in the same format as before?

Reply:
Whatever `CellWritableComparable.write()`writes must be compatible with `CellWritableComparable.readFields`
which is currently not the case (see [this exception in HBASE-25839](https://issues.apache.org/jira/browse/HBASE-25839?focusedCommentId=17338869&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17338869). The phrase "keyLength=0" inside the exception message comes actually from the statement `out.writeIn(0)`in line 139).

My proposal aligns the "write" with the existing `CellWritableComparable.readFields`. Theoretically one can adjust `CellWritableComparable.readFields` so that it becomes compatible with the current `write` method but this looks more complicated to me.

AFAICS the key is only used during the map sorting phase. The [reducer](https://github.com/apache/hbase/blob/f76a601273e834267b55c0cda12474590283fd4c/hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mapreduce/Import.java#L267) doesn't care about the key at all.

Reply:
You are basically reverting changes from commit 0a24178d06ac3dd5780522d52b5054e7aff1e50f from 4 years ago (HBASE-18649). Can you get rid of that exception if you just remove the `out.writeInt(0)` there? I'm not sure, though, on other potential side effects. Maybe we would also need to change `Import.readFields` to stay aligned with HBASE-18649 (I guess we could use `CellBuilder` to construct a `Cell` directly).

Reply:
I tried this, but it doesn't work unfortunately. If you look at the `readFields` method you'll see that it calls `KeyValue.create(DataInput in)` which in turn calls [`KeyValue.create(int length, DataInput in)`](https://github.com/apache/hbase/blob/6cfff27465620112eb308cf881e6321d928743ed/hbase-common/src/main/java/org/apache/hadoop/hbase/KeyValue.java#L2281). The length is read as integer from the DataInput which kind of matches what `write` has written before in line 138 - so far so good (I'm refering to original version btw). 
Honestly I'm a little bit confused at this point and don't know whether length expects the key length or the length of the whole key-value pair. You'll see that `KeyValue` allocates a byte array of length `length`(line 2266) and so I guess that it must be the length of key+value.

Now keep in mind that the first 4 bytes are consumed from DataInput (the length) and `KeyValue.create` calls:
`in.readFully(bytes); return new KeyValue(bytes, 0, length);` - thus `bytes` starts with four '0' bytes coming from `Import.write line 139: out.writeInt(0);`!

The [KeyValue constructure](https://github.com/apache/hbase/blob/6cfff27465620112eb308cf881e6321d928743ed/hbase-common/src/main/java/org/apache/hadoop/hbase/KeyValue.java#L343) in turn calls [`KeyValueUtil.checkKeyValueBytes`](https://github.com/apache/hbase/blob/6cfff27465620112eb308cf881e6321d928743ed/hbase-common/src/main/java/org/apache/hadoop/hbase/KeyValueUtil.java#L520). And here I discovered that the serialization can never match the deserialization because `KeyValueUtil.checkKeyValueBytes` expects the key len at the beginning of the `bytes`buffer (note that `offset==0`) but this '0' as mentioned before.

To sum it up:
`CellWritableComparable.write`writes:
[ [4 bytes key length], [4 bytes '0'], [...] ]

`CellWritableComparable.readFields` expects:
[ [4bytes key-value length], [4 bytes key length], [...] ]

Honestly, I wonder how that ever worked ;-)

Reply:
Te patch in HBASE-18649 changed this part and made to write only the key part of the cell (rowkey + cf + q + ts + type) instead of the entire KV.  So it writes key length
out.writeInt(PrivateCellUtil.estimatedSerializedSizeOfKey(kv));
And then value len as 0
and then key part.
But this is not compatible with counter part read where it expects a prefix int part which says the overall length of this KV (Ya what we wrote is still a KV , or call it Cell, where value is empty).   
So I agree to the fact that this never worked. 
We can not get rid of out.writeInt(0).
On the other hand we have to write this entire KV length what we are going to write.  This is key len + value len(0) + 4 + 4  (we write these extra 4 bytes twice for these 2 lengths too)

int keyLen = CellUtil.estimatedSerializedSizeOfKey(kv);
int valueLen = 0; // We avoid writing value here. So just serialize as if an empty value.
out.writeInt(keyLen+ valueLen + KeyValue.KEYVALUE_INFRASTRUCTURE_SIZE);
out.writeInt(keyLen);
out.writeInt(valueLen);
CellUtil.writeFlatKey(kv, out);

Not sure whether we already have some util method for this.  Else worth adding?